### DISCUSSION TITLE
Open Source + GitHub + Git Bootcamp what should be included... ?

### DISCUSSION THREAD

**Discussion by eddiejaoude** [EddieHubCommunity/support#434] [Ideas]
I am planning to do a Open Source, GitHub and Git crash course, what do you think should be included?

## Thoughts

- Target audience
- Discord channel for students only (standup, achievements, questions)
- Limit the monthly intake (100?)
- Each module to have actionable steps of tasks to complete
- Inform with statistics

## Course outline

- Introduction (public / free)
   - What the course contains
   - What you will get out of it
   - Who the course is for
   - Share some tips ...
   - If you want to buy the course to not, it is free to join our community on GitHub Org and Discord
- Glossary of technical terms (for example: Pull Request === PR)
- What, Why, How is Open Source
   - similar to instagram, like, comment, share but can also contribute and make improvements
- Benefits of being involved in Open Source
   - Learning faster
   - Improving communication skills
   - Grow a network
   - How you can make money from Open Source
      - Jobs
      - Freelance
   - Being new to a project / technology can add so much value to it
- What is GitHub
- Setting up your GitHub profile for success
   - Create an account, I recommend using your name, (e.g. `eddiejaoude`), this will make up your url and future employers will see this
   - You can change it later, but it is not perfect
   - Set your profile picture, bio, link
   - Customise your GitHub profile by creating a repo with the same name as your account name, even if using the default template
   - Pin the repo
- Git repo
   - Commit
   - Release
- Git
   - what is git
   - clone
   - stash
   - stage / unstage
   - commit
   - writing a good commit message
   - remotes 
   - pull / fetch / push
   - rebase
   - branching
   - resolving conflicts
   - tags
- Issues
   - Labels
   - Milestones
   - Task list
- Project boards
- Pull Requests
   - How to get your PRs accepted
   - What not to do (tag / DM the maintainers that you raised a PR)
   - Demand it is merged
- What is CI/CD
- Actions
   - Use existing Actions
   - Creating your own custom Actions
- Markdown
- Contributing to Open Source
   - How to find Open Source projects
   - Open Source etiquette
      - how long to have an issue assigned
      - how long to have a PR open for
   - How to create a great PR
      - Draft PRs
      - Task list
   - How to create a great README
- How to customise your GitHub profile
   - examples
- Project maintainer
   - Open Source Licensing 
   - Contributors file
   - Issue template
   - PR template
   - How to make your project inclusive (insights -> community tab) 
   - how to review a PR
   - Writing a good README
- Wiki vs `docs/`
- Settings page
   - Enable / disable settings (e.g. wiki)
   - Personal Access tokens
   - ...
- GitHub Pages (free static hosting)
- GitHub Webhooks
- GitHub CLI
- CodeSpaces
    - ngrok to make it publicly available
- Repo best practices
   - label automation with GH Action
   - docs automation with AsciiDoctor and GH Action
   - stale GH Action
   - Husky
   - package with Docker
   - create release with changelog
- guests to talk about open source
   - maintainers
   - ...

Briefly touch on
- Freelancing
- Portfolio

**Comment by phazonoverload** [comment]
Something about open source etiquette would be great.

**Reply by eddiejaoude** [reply]
Oh yes, love that 👍 . Thank you!

**Reply by eddiejaoude** [reply]
I updated and added it to the list above 👆

**Reply by Emelda-GA** [reply]
This is very helpful to us newbies. Thanks.

**Reply by eddiejaoude** [reply]
Thank you for the feedback and support 👍

**Reply by T9ipa** [reply]
I've done it right! Our days are harsh!

**Comment by Ruth-ikegah** [comment]
How to find open source projects is an addition from me

**Reply by eddiejaoude** [reply]
Great idea, thank you 👍

**Reply by eddiejaoude** [reply]
I updated and added it to the list above 👆

**Comment by RasPiPkr** [comment]
You could add licence and contribution into it, and don't forget the readme

**Reply by eddiejaoude** [reply]
OMG YES!! Thank you 👍

**Reply by eddiejaoude** [reply]
I updated and added it to the list above 👆

**Reply by RasPiPkr** [reply]
:thumbsup: can't wait

**Comment by vickywane** [comment]
Git rebasing as standard practice to apply when contributing to Open Sourced Projects. 

- How, When, and Why you should rebase git commits.
- How to use the interactive rebasing editor.

**Reply by eddiejaoude** [reply]
Great idea, I have added it to the list in the description above 👍  Thanks 🤓

**Comment by jebitok-dev** [comment]
How to find easy or beginner-friendly issues on large open source projects that has many contributors and ethics of contribute (in terms of time limit incase I've an assigned to for how long should I stay without contributing to it)

**Reply by eddiejaoude** [reply]
Great idea, I have added it to the list in the description above 👍 Thanks 🤓

**Reply by Kitecatcher** [reply]
Hey......... I know what I'm doing on this site yet but I do have goals and ambitions. a where to start would be great! I just want to get my feet wet and if that means tossed in with the sharks. Then so be it that's the best way to learn! super interested in robotics, python, 3d printing artificial intelligence and so much more.  if you need a beginner to test or give bug reports let me know!

**Comment by Shababeu** [comment]
1. Jenkins integration with GitHub.
2. Tagging,Branching and Merging.

Thank You...

**Reply by eddiejaoude** [reply]
Tagging, Branching, Merging are great ideas, I have added it to the list in the description above 👍  Thanks 🤓 

I will include GitHub Actions for CI, instead of Jenkins

**Reply by Sam-Rowe** [reply]
👍 for branching! 
There are so many branch strategies this could be a whole section on it's own but for this doc maybe just describe one?

**Reply by eddiejaoude** [reply]
What are your favourite / recommended strategies? I think gitflow is the most popular still?

**Reply by Sam-Rowe** [reply]
I can only apologise for picking up old threads a couple of years later but here goes.

GitFlow is popular in Enterprises, but it really isn't great. The complexity of merging is not something I like. https://nvie.com/posts/a-successful-git-branching-model/ 

However GitHub flow is simple and easy to follow.
https://docs.github.com/en/get-started/quickstart/github-flow

and I see big orgs/projects make great use of Release Flow https://devblogs.microsoft.com/devops/release-flow-how-we-do-branching-on-the-vsts-team/

**Reply by eddiejaoude** [reply]
Thanks Sam, those are some great links!

**Comment by aniketsingh98571** [comment]
@eddiejaoude Profile Building using new Github Readme.MD Feature.

**Reply by eddiejaoude** [reply]
Great idea, I have added it to the list in the description above 👍  Thanks 🤓

**Comment by wbrijesh** [comment]
Encouraging, respecting diversity in open source

**Reply by eddiejaoude** [reply]
Great idea, I have added it to the list in the description above 👍  Thanks 🤓

**Comment by wbrijesh** [comment]
And also contributor agreement (cla)

**Reply by eddiejaoude** [reply]
Great idea, I have added it to the list in the description above 👍  Thanks 🤓

**Comment by frankiefab100** [comment]
MARKDOWN crash course

**Reply by eddiejaoude** [reply]
Great idea, I have added it to the list in the description above 👍  Thanks 🤓

**Comment by Sam-Rowe** [comment]
I found Git stash useful when I was less confident in my contributions. It helped me rapidly test ideas on trunk branch before I had confidence to create a feature branch and PR back in.

**Reply by eddiejaoude** [reply]
Thanks @Sam-Rowe 👍

**Comment by HarshKapadia2** [comment]
Whenever I give a talk on Git and GitHub, I include the following points:
https://www.github.com/HarshKapadia2/talks/tree/main/git_basics%2Fcontent.md

**Reply by eddiejaoude** [reply]
Thank you for sharing 👍

**Comment by Devansu-Yadav** [comment]
@eddiejaoude I would like to know about how we can work with the codebase of large repositories of different organizations and how we can understand their codebase.

**Reply by eddiejaoude** [reply]
Thank you, I will have a think of how to cover this 👍

**Comment by mikeysan** [comment]
A "Who is this for" section. Target audience.
Yes, I can hear you saying 'anyone'. :) But I mean from a technical standpoint. What would you consider the minumum I would need to know in other to benefit from your awesome knowledge?

**Reply by eddiejaoude** [reply]
Great point!! I will make a note and having a think about this 👍

**Comment by Akshu-on-github** [comment]
What about GitHub's hosting of static websites? The `username.github.io` and the `project.github.io` ones would be really useful for an online resume

**Reply by eddiejaoude** [reply]
Oh definitely YES 👍 . Thank you 🤓

**Comment by mstanciu552** [comment]
How about a section with how to learn about the preferences of project mantainers by reading the CONTRIBUTING.md if there is one or how to figure out code preferences from the actual code.
I mean how to not dive straight into work and then have your PR rejected because your code doesn't match the preferences of the manager.
I hope I made myself clear about the info I would like to know. If not sorry :yum:.

**Reply by eddiejaoude** [reply]
this is a great topic, thank you 👍

**Comment by Vyvy-vi** [comment]
With the addition to creating a README, how to create an **engaging** README ?
(personally, even though I tried designing my README to look good, a reasonable share of contributors end up not reading it)

**Reply by eddiejaoude** [reply]
yes 👍  - I have seen some good examples I can use, thank you

**Comment by mikeysan** [comment]
New tool, GitHub Cli (gh). It might be good to include a section on it as well.

**Reply by eddiejaoude** [reply]
Oh yes!! Great idea 👍 Thank you!

**Comment by BOLT04** [comment]
What does everyone think about [GitHub Webhooks](https://docs.github.com/en/free-pro-team@latest/developers/webhooks-and-events/about-webhooks). I think it can be used for cool integrations like Discord and much more, so it would be cool to cover and maybe even build something with them (or just show an easier integration) 😃.

**Reply by eddiejaoude** [reply]
Oh just seeing this, I must have missed the notification 🤦‍♂️ 

I used to use webhooks a lot, but now I find I use GitHub Actions more. Webhooks are great though! I think you are right, I should include them. Thank you 🤓

**Comment by manavmodi22** [comment]
Invite open Source Project Maintainers to talk about their projects and how Open Source has played a vital role in shaping the project.

**Reply by eddiejaoude** [reply]
oh interesting idea 👍 thank you!

**Comment by CBID2** [comment]
Hi @eddiejaoude! :) I think mentioning PR that doesn’t involve using Git can be helpful, especially for those who are uncomfortable with using Git. Maybe add some homework assignments to reinforce learning. Furthermore, I think you should consider making this course free. Breaking into Tech is already a costly feat for some folks, and making it free would increase accessibility. This is something that I learned from observing freeCodeCamp.

**Reply by eddiejaoude** [reply]
Thank you for the great ideas 👍 

I already have a lot of free content on my YouTube channel (weekly videos and live streams), and my other current course on customising our GitHub profiles is complete free also http://eddiejaoude.io/course-github-profile-landing

**Reply by CBID2** [reply]
Oh I see. Maybe you can add an Open Source cheatsheet guide as a freebie?

**Reply by eddiejaoude** [reply]
Sure,  I want to do more cheatsheets 👍

**Reply by CBID2** [reply]
It could be posted on your YouTube channel too! :)

**Comment by Kevinkp09** [comment]
Is this crash course completed or its still in work? If its out there i need to watch and learn.

**Reply by eddiejaoude** [reply]
I have done various videos on this on my YouTube channel http://linkfree.io/eddiejaoude

**Reply by Kevinkp09** [reply]
Thanks Eddie. You are the real OG
